internal/filter/filter.go

```diff
@@ -177,12 +184,8 @@ func List(patterns []string, str string) (matched bool, childMayMatch bool, err
 			return false, false, err
 		}
 
-		matched = matched || m
-		childMayMatch = childMayMatch || c
-
-		if matched && childMayMatch {
-			return true, true, nil
-		}
+		matched = matched && !m || m && !negate
+		childMayMatch = childMayMatch && !(m && c) || c && !negate
```

Comment:
If I understand `match` and `childMatch` correctly, `m==true` implies `c==true`. That would simplify the last line to `childMayMatch && !m || c && !negate`.

That would lead to a simpler explanation for this bit magic: the matched/childMayMatch flag is reset if the pattern fully matches, it is set again if this is a not-negated pattern and match/childMatch were successful.

It took me a rather long time to fully understand the effects of these Boolean expressions and why these are correct. I've tried to find a formula that's more intuitive to understand, but I'm not sure whether the result is actually better:
```
matched = !negate && (matched || m) || (negate && matched && !m)
childMayMatch = !negate && (childMayMatch || c) || (negate && childMayMatch && !m)
```

So I think these two lines need a comment at the very least.

Reply:
While your `matched` formula is equivalent, the `childMayMatch` is not. Are the tests successful with your version?

Reply:
Yes the tests are successful. In case `m==true implies c==true` is correct, this simplifies `m && c` to `m` and then you end up with my variant of the `childMayMatch` formula.

Reply:
Oh, OK, I didn't catch the implication. Fine by me.

Reply:
Which variant of the matched, childMayMatch expressions do you find more readable? Another option would be add an if-else-block for the normal and negated version. That would be a few more lines of code, but probably the easiest to understand variant.

Reply:
I don't have a strong opinion about this. The two options would be:

```go
matched = matched && !m || m && !negate
childMayMatch = childMayMatch && !m || c && !negate
```

Or

```go
if negate {
   matched = matched && !m
   childMayMatch = childMayMatch && !m
} else {
   matched = matched || m
   childMayMatch = childMayMatch || c
}
```

If it is correct, the latest form has some "logical look" in it. So I would go with it.

Reply:
Thanks @vincentbernat. I agree with @MichaelEischer in finding the if-else formulation easier to read.